File: threaded/primesthreaded.rb
#!/usr/bin/env ruby

require 'thread'
@mutex = Mutex.new
@no_primes = 0

def isPrime(num)
    return false if (num <= 1)
    return true if (num <= 3)
    return false if (((num % 2) == 0) or ((num % 3) == 0))
    c = 5
    while c*c <= num
        return false if (((num % c) == 0) or ((num % (c+2)) == 0))
        c+=6
    end
    return true      
end

def countPrimes(min, max, step)
    num = min
    while num <= max
        if isPrime(num)
            @mutex.synchronize {
                @no_primes += 1
            }
        end
        num = num + step
    end
end    


min = 0
max = 20000000

puts "Ruby 4 threads"

beginning_time = Time.now

threads = []
4.times do |i|
   threads[i] = Thread.new { countPrimes(min+i, max, 4) }
end

threads.map(&:join)

@mutex.lock

end_time = Time.now
delta_time = end_time - beginning_time

puts "Number of primes in the interval [#{min},#{max}]: #{@no_primes}"
puts "Time elapsed: #{delta_time} seconds"

File: threaded/primes_threaded.c
#include <stdio.h>
#include <pthread.h>
#include <time.h>

void * thread_sum(void *);
int isPrime(int num);
long diff_nano(struct timespec *start, struct timespec *end);

const int nThreads = 4;
pthread_mutex_t mVar = PTHREAD_MUTEX_INITIALIZER;
int TotalCount = 0;
int minprime = 0, maxprime = 20000000;

void *thread_sum(void *no)
{
    int *ifrom, iCount;
    ifrom = (int*) no;
    for (iCount = *ifrom; iCount <= maxprime; iCount+=(nThreads))
    {
        if ( isPrime(iCount) == 1)
        {
            pthread_mutex_lock(&mVar);
            TotalCount++;
            pthread_mutex_unlock(&mVar);
        }
    }
    pthread_exit(NULL);
}

int isPrime(int num)
{
    if (num <= 1) return 0;
    if (num <= 3) return 1;
    if (((num % 2) == 0) || ((num % 3) == 0)) return 0;
    int c;
    for (c = 5; c * c <= num; c += 6)
        if (((num % c) == 0) || ((num % (c + 2)) == 0)) return 0;
    return 1;
}

long diff_nano(struct timespec *start, struct timespec *end)
{
    /* ns */
    return ( (end->tv_sec * 1000000000 ) + (end->tv_nsec) -
             (start->tv_sec * 1000000000) - (start->tv_nsec) );
}

int main()
{
    int iCount;
    int threadPara[nThreads];
    
    struct timespec tstart, tend;
    pthread_t tid[nThreads];
		
    printf("C with %d Threads\n",nThreads);

    clock_gettime(CLOCK_MONOTONIC, &tstart);

    for (iCount = 0; iCount < nThreads; iCount++)
    {
        threadPara[iCount] = iCount;
        pthread_create(&tid[iCount], NULL, &thread_sum, (void *) &threadPara[iCount]);
    }

    for (iCount = 0; iCount < nThreads; iCount++)
    {
        pthread_join(tid[iCount], NULL);
    }

    clock_gettime(CLOCK_MONOTONIC, &tend);
    double time_spent = diff_nano(&tstart, &tend) / 1000000000.;

    printf("Number of primes in the interval [%d,%d]: %d\n",
           minprime, maxprime, TotalCount);
    printf("Time elapsed: %f seconds\n",time_spent);
         
    return 0;
}
